tests/dragonfly/cluster_mgr_test.py

```diff
@@ -0,0 +1,162 @@
+import subprocess
+import redis
+from redis import asyncio as aioredis
+from .utility import *
+from . import dfly_args
+
+BASE_PORT = 40001
+
+
+async def insert_cluster_data(cluster_client: redis.RedisCluster):
+    for i in range(1_000):
+        await cluster_client.set(i, i)
+
+
+async def check_cluster_data(cluster_client: redis.RedisCluster):
+    for i in range(1_000):
+        assert await cluster_client.get(i) == str(i)
+
+
+def run_cluster_mgr(args):
+    result = subprocess.run(["../tools/cluster_mgr.py", *args])
+    return result.returncode == 0
+
+
+@dfly_args({"proactor_threads": 2, "cluster_mode": "yes"})
+async def test_cluster_mgr(df_local_factory):
+    NODES = 3
+    masters = [df_local_factory.create(port=BASE_PORT + i) for i in range(NODES)]
+    replicas = [df_local_factory.create(port=BASE_PORT + 100 + i) for i in range(NODES)]
+    df_local_factory.start_all([*masters, *replicas])
+
+    # Initialize a cluster (all slots belong to node 0)
+    assert run_cluster_mgr(["--action=config_single_remote", f"--target_port={BASE_PORT}"])
+    for i in range(1, NODES):
+        assert run_cluster_mgr(
+            ["--action=attach", f"--target_port={BASE_PORT}", f"--attach_port={BASE_PORT+i}"]
+        )
+
+    # Feed the cluster with data and test that it works correctly
+    client = aioredis.RedisCluster(decode_responses=True, host="127.0.0.1", port=masters[0].port)
+    await insert_cluster_data(client)
+    await check_cluster_data(client)
+
+    # Migrate ~half of the slots to node 1
+    assert run_cluster_mgr(
+        [
+            f"--action=migrate",
+            f"--target_port={BASE_PORT + 1}",
+            f"--slot_start=8000",
+            f"--slot_end=16383",
+        ]
+    )
+    await check_cluster_data(client)
+
+    # Can only detach node 2 (with no assigned slots)
+    assert not run_cluster_mgr(["--action=detach", f"--target_port={BASE_PORT}"])
+    assert not run_cluster_mgr(["--action=detach", f"--target_port={BASE_PORT + 1}"])
+    assert run_cluster_mgr(["--action=detach", f"--target_port={BASE_PORT + 2}"])
+    await check_cluster_data(client)
+
+    # Can't attach non-replica as replica
+    assert not run_cluster_mgr(
+        [
+            f"--action=attach",
+            f"--target_port={BASE_PORT}",
+            f"--attach_port={BASE_PORT+2}",
+            f"--attach_as_replica=True",
+        ]
+    )
+
+    # Reattach node 2 and migrate some slots to it
+    assert run_cluster_mgr(
+        ["--action=attach", f"--target_port={BASE_PORT}", f"--attach_port={BASE_PORT+2}"]
+    )
+    await check_cluster_data(client)
+    # Can't migrate from 2 nodes
```

Comment:
what do you mean? why cant we migrate from 2 nodes?

Reply:
slots 7k-8k belong to node0, while 8k-9k belong to node1
migrating of the entire range in cluster_mgr (in a single command) is unsupported
i updated the comment